Conflicting traits in protocol dependency at codegen time

I've seen reports of people getting issues when trying to compile a project in sbt, when that project has a codegen-time dependency on a version of smithy4s-protocol that doesn't match the one of the sbt plugin.

## How to reproduce

1. Make an sbt project like this:

```scala
//project/plugins.sbt
addSbtPlugin("com.disneystreaming.smithy4s" % "smithy4s-sbt-codegen" % "0.17.4")
```

```scala
//build.sbt
import smithy4s.codegen.Smithy4sCodegenPlugin

ThisBuild / scalaVersion := "2.13.10"

Global / onChangedBuildSource := ReloadOnSourceChanges

// imagine this is published externally
lazy val lib = project
  .settings(
    autoScalaLibrary := false,
    crossPaths := false,
    libraryDependencies ++= Seq(
      "com.disneystreaming.smithy4s" % "smithy4s-protocol" % "0.17.2"
    )
  )

lazy val main = project
  .enablePlugins(Smithy4sCodegenPlugin)
  .dependsOn(lib)

lazy val root = project
  .in(file("."))
```

You can find a full project like this [here](https://github.com/kubukoz/demos/tree/smithy4s-conflict-protocol).

3. Run `sbt main/smithy4sCodegen` in it

4. See something like this:

```
...
[error] stack trace is suppressed; run last main / Compile / smithy4sCodegen for the full output
[error] (main / Compile / smithy4sCodegen) software.amazon.smithy.model.validation.ValidatedResultException: Result contained ERROR severity validation events: 
[error] [ERROR] smithy4s.meta#Classpath: Conflicting `smithy.api#documentation` trait found on shape `smithy4s.meta#Classpath`. The previous trait was defined at `jar:file:/Users/kubukoz/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/com/disneystreaming/smithy4s/smithy4s-protocol/0.17.2/smithy4s-protocol-0.17.2.jar!/META-INF/smithy/smithy4s.meta.smithy [87, 1]`, and a conflicting trait was defined at `jar:file:/Users/kubukoz/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/com/disneystreaming/smithy4s/smithy4s-protocol/0.17.4/smithy4s-protocol-0.17.4.jar!/META-INF/smithy/smithy4s.meta.smithy [93, 1]`. | Model jar:file:/Users/kubukoz/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/com/disneystreaming/smithy4s/smithy4s-protocol/0.17.4/smithy4s-protocol-0.17.4.jar!/META-INF/smithy/smithy4s.meta.smithy:93:1
[error] [ERROR] smithy4s.meta#Classpath: Conflicting `smithy.api#pattern` trait found on shape `smithy4s.meta#Classpath`. The previous trait was defined at `jar:file:/Users/kubukoz/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/com/disneystreaming/smithy4s/smithy4s-protocol/0.17.2/smithy4s-protocol-0.17.2.jar!/META-INF/smithy/smithy4s.meta.smithy [88, 1]`, and a conflicting trait was defined at `jar:file:/Users/kubukoz/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/com/disneystreaming/smithy4s/smithy4s-protocol/0.17.4/smithy4s-protocol-0.17.4.jar!/META-INF/smithy/smithy4s.meta.smithy [95, 1]`. | Model jar:file:/Users/kubukoz/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/com/disneystreaming/smithy4s/smithy4s-protocol/0.17.4/smithy4s-protocol-0.17.4.jar!/META-INF/smithy/smithy4s.meta.smithy:95:1
[error] [ERROR] smithy4s.meta#Import: Conflicting `smithy.api#documentation` trait found on shape `smithy4s.meta#Import`. The previous trait was defined at `jar:file:/Users/kubukoz/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/com/disneystreaming/smithy4s/smithy4s-protocol/0.17.2/smithy4s-protocol-0.17.2.jar!/META-INF/smithy/smithy4s.meta.smithy [91, 1]`, and a conflicting trait was defined at `jar:file:/Users/kubukoz/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/com/disneystreaming/smithy4s/smithy4s-protocol/0.17.4/smithy4s-protocol-0.17.4.jar!/META-INF/smithy/smithy4s.meta.smithy [98, 1]`. | Model jar:file:/Users/kubukoz/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/com/disneystreaming/smithy4s/smithy4s-protocol/0.17.4/smithy4s-protocol-0.17.4.jar!/META-INF/smithy/smithy4s.meta.smithy:98:1
[error] [ERROR] smithy4s.meta#Import: Conflicting `smithy.api#pattern` trait found on shape `smithy4s.meta#Import`. The previous trait was defined at `jar:file:/Users/kubukoz/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/com/disneystreaming/smithy4s/smithy4s-protocol/0.17.2/smithy4s-protocol-0.17.2.jar!/META-INF/smithy/smithy4s.meta.smithy [92, 1]`, and a conflicting trait was defined at `jar:file:/Users/kubukoz/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/com/disneystreaming/smithy4s/smithy4s-protocol/0.17.4/smithy4s-protocol-0.17.4.jar!/META-INF/smithy/smithy4s.meta.smithy [100, 1]`. | Model jar:file:/Users/kubukoz/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/com/disneystreaming/smithy4s/smithy4s-protocol/0.17.4/smithy4s-protocol-0.17.4.jar!/META-INF/smithy/smithy4s.meta.smithy:100:1
[error] Total time: 0 s, completed Feb 27, 2023, 12:56:10 PM
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -7,7 +7,7 @@ Thank you!
 
 # 0.18.32
 
-* Fixes an issue in which smithy4s-protocol would conflict with its previous versions during codegen ([#837](https://github.com/disneystreaming/smithy4s/issues/837)) in [#1672](https://github.com/disneystreaming/smithy4s/pull/1672)
+* codegen: Fix an issue in which smithy4s-protocol would conflict with its previous versions if they're in the dependencies ([#837](https://github.com/disneystreaming/smithy4s/issues/837)) in [#1672](https://github.com/disneystreaming/smithy4s/pull/1672)
 * Adds a missing space in the type annotations of smart constructors in [#1674](https://github.com/disneystreaming/smithy4s/pull/1674)
 * More expressive namespace patterns for the codegen in [#1649](https://github.com/disneystreaming/smithy4s/pull/1649)
 

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -7,6 +7,7 @@ Thank you!
 
 # 0.18.32
 
+* Fixes an issue in which smithy4s-protocol would conflict with its previous versions during codegen ([#837](https://github.com/disneystreaming/smithy4s/issues/837)) in [#1672](https://github.com/disneystreaming/smithy4s/pull/1672)
 * Adds a missing space in the type annotations of smart constructors in [#1674](https://github.com/disneystreaming/smithy4s/pull/1674)
 * More expressive namespace patterns for the codegen in [#1649](https://github.com/disneystreaming/smithy4s/pull/1649)
 

--- a/build.sbt
+++ b/build.sbt
@@ -495,7 +495,7 @@ lazy val codegenPlugin = (projectMatrix in file("modules/codegen-plugin"))
 
         // for sbt
         (codegen.jvm(Scala212) / publishLocal).value,
-        (protocol.jvm(autoScalaLibrary = false) / publishLocal).value
+        (protocolJvm / publishLocal).value
       )
       publishLocal.value
     },
@@ -536,7 +536,7 @@ lazy val millCodegenPlugin = projectMatrix
         (codegen.jvm(Scala213) / publishLocal).value,
 
         // for mill
-        (protocol.jvm(autoScalaLibrary = false) / publishLocal).value
+        (protocolJvm / publishLocal).value
       )
       publishLocal.value
     },
@@ -589,6 +589,8 @@ lazy val protocol = projectMatrix
     libraryDependencies += Dependencies.Smithy.model,
     javacOptions ++= Seq("--release", "8")
   )
+
+lazy val protocolJvm = protocol.jvm(autoScalaLibrary = false)
 
 lazy val protocolTests = projectMatrix
   .in(file("modules/protocol-tests"))

--- a/build.sbt
+++ b/build.sbt
@@ -409,7 +409,7 @@ lazy val codegen = projectMatrix
       "alloyOrg" -> Dependencies.Alloy.org,
       "alloyVersion" -> Dependencies.Alloy.alloyVersion,
       "smithy4sOrg" -> organization.value,
-      "protocolArtifactName" -> "smithy4s-protocol",
+      "protocolArtifactName" -> "smithy4s-protocol"
     ),
     buildInfoPackage := "smithy4s.codegen",
     libraryDependencies ++= Seq(
@@ -435,7 +435,10 @@ lazy val codegen = projectMatrix
       sourceManaged
         .map(AwsBoilerplate.generate(_))
         .taskValue,
-    }
+    },
+    (Compile / compile) := (Compile / compile)
+      .dependsOn((protocol.jvm(autoScalaLibrary = false) / publishLocal))
+      .value
   )
 
 /**

--- a/modules/codegen/src/smithy4s/codegen/SmithyBuildJson.scala
+++ b/modules/codegen/src/smithy4s/codegen/SmithyBuildJson.scala
@@ -26,9 +26,6 @@ import scala.collection.immutable.ListSet
 
 private[codegen] object SmithyBuildJson {
 
-  val protocolDependency =
-    s"${BuildInfo.smithy4sOrg}:${BuildInfo.protocolArtifactName}:${BuildInfo.version}"
-
   def toJson(
       sources: ListSet[String],
       dependencies: ListSet[String],

--- a/modules/codegen/src/smithy4s/codegen/internals/ModelLoader.scala
+++ b/modules/codegen/src/smithy4s/codegen/internals/ModelLoader.scala
@@ -47,7 +47,7 @@ private[codegen] object ModelLoader {
   ): (ClassLoader, Model) = {
     val currentClassLoader = this.getClass().getClassLoader()
     val deps = resolveDependencies(
-      dependencies :+ s"${BuildInfo.smithy4sOrg}:${BuildInfo.protocolArtifactName}:${BuildInfo.version}",
+      dependencies :+ protocolDependency,
       localJars,
       repositories
     )

--- a/modules/codegen/src/smithy4s/codegen/internals/ModelLoader.scala
+++ b/modules/codegen/src/smithy4s/codegen/internals/ModelLoader.scala
@@ -46,7 +46,11 @@ private[codegen] object ModelLoader {
       localJars: List[os.Path]
   ): (ClassLoader, Model) = {
     val currentClassLoader = this.getClass().getClassLoader()
-    val deps = resolveDependencies(dependencies, localJars, repositories)
+    val deps = resolveDependencies(
+      dependencies :+ s"${BuildInfo.smithy4sOrg}:${BuildInfo.protocolArtifactName}:${BuildInfo.version}",
+      localJars,
+      repositories
+    )
 
     val modelsInJars = deps.flatMap { file =>
       Using.resource(
@@ -178,13 +182,9 @@ private[codegen] object ModelLoader {
         classLoader: ClassLoader,
         discoverModels: Boolean
     ): ModelAssembler = {
-      val smithy4sResources = List(
-        "META-INF/smithy/smithy4s.meta.smithy"
-      ).map(classLoader.getResource)
-
       if (discoverModels) {
         assembler.discoverModels(classLoader)
-      } else addImports(smithy4sResources)
+      } else assembler
     }
   }
 

--- a/modules/codegen/src/smithy4s/codegen/package.scala
+++ b/modules/codegen/src/smithy4s/codegen/package.scala
@@ -17,5 +17,9 @@
 package smithy4s
 
 package object codegen {
-  val SMITHY4S_DEPENDENCIES = "smithy4sDependencies"
+  val SMITHY4S_DEPENDENCIES: String = "smithy4sDependencies"
+
+  val protocolDependency: String =
+    s"${BuildInfo.smithy4sOrg}:${BuildInfo.protocolArtifactName}:${BuildInfo.version}"
+
 }

--- a/modules/codegen/test/src/smithy4s/codegen/internals/ModelLoaderSpec.scala
+++ b/modules/codegen/test/src/smithy4s/codegen/internals/ModelLoaderSpec.scala
@@ -71,6 +71,16 @@ class ModelLoaderSpec extends FunSuite {
     // nothing failed
   }
 
+  test(
+    "ModelLoader can load a version of Alloy conflicting against the current"
+  ) {
+    doLoad(
+      dependencies = List("com.disneystreaming.alloy:alloy-core:0.1.18"),
+      repositories = Nil
+    )
+    // nothing failed
+  }
+
   test("ModelLoader can load a dependency from s01 if it has a + in the name") {
     val model = doLoad(
       dependencies =

--- a/modules/codegen/test/src/smithy4s/codegen/internals/ModelLoaderSpec.scala
+++ b/modules/codegen/test/src/smithy4s/codegen/internals/ModelLoaderSpec.scala
@@ -60,6 +60,17 @@ class ModelLoaderSpec extends FunSuite {
     model.expectShape(ShapeId.from("testlibrary#MyString"))
   }
 
+  test(
+    "ModelLoader can load a version of the smithy4s protocol conflicting against the current"
+  ) {
+    doLoad(
+      dependencies =
+        List("com.disneystreaming.smithy4s:smithy4s-protocol:0.18.29"),
+      repositories = Nil
+    )
+    // nothing failed
+  }
+
   test("ModelLoader can load a dependency from s01 if it has a + in the name") {
     val model = doLoad(
       dependencies =

